libraries/AP_ExternalAHRS/AP_ExternalAHRS_AdvancedNavigation.h

```diff
@@ -0,0 +1,378 @@
+        float heading;
+        float tilt_sd;
+        float heading_sd;
+        union {
+            uint16_t r;
+            struct {
+                uint16_t fix_type :3;
+                uint16_t velocity_valid :1;
+                uint16_t time_valid :1;
+                uint16_t external_gnss :1;
+                uint16_t tilt_valid :1; /* This field will only be valid if an external dual antenna GNSS system is connected */
+                uint16_t heading_valid :1; /* This field will only be valid if an external dual antenna GNSS system is connected */
+            } b;
+        } flags;
+    };
+
+    struct PACKED AN_SATELLITES {
+        float hdop;
+        float vdop;
+        uint8_t gps_satellites;
+        uint8_t glonass_satellites;
+        uint8_t beidou_satellites;
+        uint8_t galileo_satellites;
+        uint8_t sbas_satellites;
+    } *_last_satellites;
+
+    struct PACKED AN_PERIOD {
+        uint8_t id;
+        uint32_t packet_period;
+    };
+
+    struct PACKED AN_PACKETS_PERIOD {
+        uint8_t permanent;
+        uint8_t clear_existing_packet_periods;
+        AN_PERIOD periods[(AN_MAXIMUM_PACKET_SIZE - 2)/sizeof(AN_PERIOD)];
+    };
+
+    class PACKED AN_PACKET
+    {
+    public:
+        uint8_t lrc;
+        uint8_t id;
+        uint8_t length;
+        uint16_t crc;
+
+        union payload {
+            uint8_t raw_packet[AN_MAXIMUM_PACKET_SIZE];
+            AN_DEVICE_INFO device_info;
+            AN_SYSTEM_STATE system_state;
+            AN_VELOCITY_STANDARD_DEVIATION velocity_standard_deviation;
+            AN_RAW_SENSORS raw_sensors;
+            AN_RAW_GNSS raw_gnss;
+            AN_SATELLITES satellites;
+            AN_PACKETS_PERIOD packets_period;
+        } payload;
+
+        void update_checks(uint8_t header_id, uint8_t header_length)
+        {
+            // Update the packet check and header id
+            crc = crc16_ccitt(payload.raw_packet, header_length, 0xFFFF);
+            id = header_id;
+
+            // Update the header LRC
+            uint8_t* id_ptr = &id;
+            lrc = ((id_ptr[0] + id_ptr[1] + id_ptr[2] + id_ptr[3]) ^ 0xFF) + 1;
+        }
+
+        uint8_t* raw_pointer()
+        {
+            return &lrc;
+        }
+
+        size_t packet_size() const
+        {
+            return AN_PACKET_HEADER_SIZE + length;
+        }
+    };
+
+    union PACKED MsgUnion {
+        MsgUnion() { }
+        uint8_t buffer[AN_PACKET_HEADER_SIZE + AN_MAXIMUM_PACKET_SIZE];
+        AN_PACKET packet;
+    } _msg;
+
+    uint16_t _current_rate;
+    uint16_t _system_status;
+    uint16_t _filter_status;
+    float _gnss_sd[3];
+
+    AP_HAL::UARTDriver *_uart;
+    HAL_Semaphore _sem;
+
+    uint32_t _baudrate;
+    int8_t _port_num;
+    bool _port_open = false;
+
+    uint32_t _last_pkt_ms;
+    uint32_t _last_state_pkt_ms;
+    uint32_t _last_device_info_pkt_ms;
+    uint32_t _last_raw_gnss_pkt_ms;
+
+    uint32_t _device_id = 0;
+    uint32_t _hardware_rev = 0;
```

Comment:
```suggestion
    uint32_t _device_id;
    uint32_t _hardware_rev;
```

these are implicitly 0 (unless you want to declare one an AdvancedNavigation devie on the stack, in which case you've got bigger problems ;-) )

We zero memory when things are `new`d in ArduPilot.